### PR TITLE
Add activation checkpointing for ViT backbones

### LINKED ISSUE #929: [FEAT] Add gradient/activation checkpointing for pretraining methods
Pretraining methods like DINOv2 and DINOv3.1 run full forward passes through all transformer blocks in the student backbone, with each block's activations held in memory for the backward pass. For
  ViT-L (24 blocks, embed_dim=1024) and ViT-G (40 blocks, embed_dim=1536), this causes high peak GPU memory usage that limits batch size and model scale on commodity hardware.

Currently there is zero usage of `torch.utils.checkpoint` anywhere in the codebase

The specific hot paths are the block iteration loops that carry gradients during training:

  - **DINOv2 ViT** (`src/lightly_train/_models/dinov2_vit/dinov2_vit_src/models/vision_transformer.py`):
    - `forward_features` (line 350): `for blk in self.blocks: x = blk(x)`
    - `forward_features_list` (line 326): `for blk in self.blocks: x = blk(x)`

  - **DINOv3 ViT** (`src/lightly_train/_models/dinov3/dinov3_src/models/vision_transformer.py`):
    - `forward_features_list` (line 264): `for _, blk in enumerate(self.blocks): x = blk(x, rope_sincos)`
    - `_get_intermediate_layers_not_chunked` (line 317): `for i, blk in enumerate(self.blocks): x = blk(x, rope_sincos)`
    
  - **ECViT** (`src/lightly_train/_models/ecvit/ecvit.py`):
    - `forward_with_grid` (line 415): `for i, blk in enumerate(self.blocks): x = blk(x, rope_sincos=rope_sincos)`

  Teacher forward passes (e.g., `DINOv2._forward_teacher` at `dinov2.py:399`, `DistillationV3._forward_teacher` at `distillationv3.py:293`) run under `@torch.no_grad()` and do not need checkpointing.

  ### Approach
  
  Add an opt-in `activation_checkpointing: bool = False` config option that wraps each transformer block call with `torch.utils.checkpoint.checkpoint(..., use_reentrant=False)` during the student forward pass.
  
cc @liopeer

### PR DESCRIPTION
**Pull request by XAheli**
Closes #929

## Summary
- Adds `ActivationCheckpointingArgs` config class with `enabled: bool` and `every_n_blocks: int`
- Applies `torch.utils.checkpoint.checkpoint(use_reentrant=False)` to block loops in DINOv2 ViT, DINOv3 ViT, and ECViT backbones
- Threads config through `pretrain()` API and CLI via `activation_checkpointing` parameter
- Auto-disables during eval/no_grad (teacher path)

## Usage
```python
  lightly_train.pretrain(
      ...,
      activation_checkpointing={"enabled": True, "every_n_blocks": 1},
  )
```
CLI: `lightly-train pretrain ... activation_checkpointing.enabled=true`

## Test plan
  - [x] 18 new tests (config, helper, 3 backbones, integration)
  - [x] 269 existing tests pass (0 regressions)
  - [x] GPU memory verified: 30.2% reduction on ViT-S/14 (706→493 MB)
  - [x] Numerical correctness: max diff = 0.00e+00
  - [x] ruff format + lint clean

### CODE REVIEW

**Comment by XAheli** [comment]
cc @liopeer

File: src/lightly_train/_commands/train.py [hunk 1]
```diff
@@ -239,6 +241,12 @@ def pretrain(
             parameter. For example, if ``model='torchvision/<model_name>'``, the
             arguments are passed to
             ``torchvision.models.get_model(model_name, **model_args)``.
+        activation_checkpointing:
```
Comment (liopeer) [inline src/lightly_train/_commands/train.py]: Would call it `activation_checkpoint_args` to make it clear that this requires a dictionary (follows convention of the rest of the API).

File: src/lightly_train/_commands/train.py [hunk 2]
```diff
@@ -369,9 +378,22 @@ def train_from_config(config: TrainConfig, called_via_train: bool = False) -> No
...
+            ac_args = validate.pydantic_model_validate(
+                ActivationCheckpointingArgs, dict(config.activation_checkpointing)
+            )
+        else:
+            ac_args = ActivationCheckpointingArgs()
+        if ac_args.enabled:
+            model_args["activation_checkpointing"] = True
+            model_args["activation_checkpointing_every_n_blocks"] = (
+                ac_args.every_n_blocks
+            )
```
Comment (liopeer) [inline src/lightly_train/_commands/train.py]: Would move this block also to `train_helpers`. And I would also validate the model against those settings, since it will only work for a subset of models and we should then throw this error early on. I think to this end you can create a lightweight Protocol for models that support those kwargs, e.g. in the `model_wrapper.py` module -> `SupportsActivationCheckpointing`.

File: src/lightly_train/_models/ecvit/ecvit.py [hunk 3]
```diff
@@ -412,8 +419,18 @@ def forward_with_grid(self, x: Tensor) -> tuple[list[Tensor], tuple[int, int]]:
         cos = cos.to(device=x_embed.device, dtype=x_embed.dtype)
         rope_sincos = sin.unsqueeze(0).unsqueeze(0), cos.unsqueeze(0).unsqueeze(0)
 
+        from lightly_train._activation_checkpointing import maybe_checkpoint
```
Comment (liopeer) [inline src/lightly_train/_models/ecvit/ecvit.py]: Would move to top of module everywhere.

**Review by liopeer — CHANGES_REQUESTED** [review]
Overall pretty clean already! ✨ 

Minor things to address, then we can merge, and continue with:
- [ ] documentation
- [ ] CHANGELOG entry

I would do them in a single PR together.

**Comment by XAheli** [comment]
@liopeer Thanks for the review! I have question on the protocol design. The `SupportsActivationCheckpointing` Protocol currently checks for `_activation_checkpointing` and `_activation_checkpointing_every_n_blocks` attributes on the backbone model. These are private attributes set in the constructors. Is this the shape you had in mind or would you prefer a different approach?

> Minor things to address, then we can merge, and continue with:
> * [ ]  documentation
> * [ ]  CHANGELOG entry
>
>  I would do them in a single PR together.

just confirming, should these be addressed in a followup PR or within this one?

**Review by liopeer — APPROVED** [review]
Sorry, I had not been super clear about what I expected from the protocol – will be more clear next time. I now actually quickly pushed it myself, thanks a lot for the great contribution!

**Comment by XAheli** [comment]
Sure no biggie!

**Comment by liopeer** [comment]
> just confirming, should these be addressed in a followup PR or within this one?

Sorry for not answering this. A follow-up for those would be great. If it's undocumented, it won't be used. 🙃
